Add 'Checksum.consume' method, wrapping chunks read from a stream

tseaver opened the pull request.

tseaver committed:
Add 'testing' extra.

tseaver committed:
Add 'Checksum.consume' for wrapping chunks read from a stream.

dhermes commented on setup.py:
```diff
@@ -78,6 +78,7 @@ def main():
             "Programming Language :: Python :: 3.7",
         ],
         cmdclass={"build_ext": BuildExtWithDLL},
+        extras_require={"testing": ["pytest", "mock"]},
```
```suggestion
        extras_require={
            "testing": ["pytest", 'mock; python_version == "2.7"']
        },
```

dhermes commented on src/crc32c/__init__.py:
```diff
@@ -95,3 +95,19 @@ def copy(self):
+
+        Args:
+            stream (readable file-like object): the stream to consume.
+            chunksize (int): the size of the read to perform
```
```suggestion
            chunksize (int): the size of the read to perform

```

dhermes commented on src/crc32c/__init__.py:
```diff
@@ -95,3 +95,19 @@ def copy(self):
+        """Consume chunks from a stream, extending our CRC32 checksum.
+
+        Args:
+            stream (readable file-like object): the stream to consume.
```
```suggestion
            stream (BinaryIO): the stream to consume.
```

@tseaver There are "generic" descriptions in https://docs.python.org/3/library/typing.html for many such types

dhermes commented on src/crc32c/__init__.py:
```diff
@@ -95,3 +95,19 @@ def copy(self):
+            stream (readable file-like object): the stream to consume.
+            chunksize (int): the size of the read to perform
+        Returns:
+            (generator) iterable of the chunks read from the stream.
```
```suggestion
            Generator[bytes, None, None]: Iterable of the chunks read from the stream.
```

dhermes commented on tests/test___init__.py:
```diff
@@ -12,6 +12,7 @@
 import functools
 import itertools
 
+import mock
```
@tseaver You should use `unittest.mock` (which means this would move to the stdlib imports section too). I usually use a [hack][1] for Python 2.7, so we can forget about it when 2020 comes:

```python
# conftest.py
if six.PY2:
    import mock

    unittest.mock = mock
    sys.modules["unittest.mock"] = unittest.mock
```

[1]: https://github.com/dhermes/bezier/blob/7c5b72e823cb8a88dbc81ea2581ba9e5fa204eba/tests/conftest.py#L27-L31

dhermes commented on tests/test___init__.py:
```diff
@@ -249,3 +250,19 @@ def test_copy():
+        helper = crc32c.Checksum()
+        expected = list(iscsi_chunks(chunksize))
+        stream = mock.Mock(spec=["read"])
+        stream.read.side_effect = expected + [""]
```
```suggestion
        stream.read.side_effect = expected + [b""]
```

dhermes commented on tests/test___init__.py:
```diff
@@ -249,3 +250,19 @@ def test_copy():
+        assert found == expected
+        for call in stream.read.call_args_list:
+            assert call[0] == (chunksize,)
+            assert call[1] == {}
```
Instead:

```python
assert call == mock.call(chunksize)
```

I'm always uneasy about indexing into `mock.call()` objects because it isn't "exhaustive" the way `==` is.

dhermes reviewed (commented):
@tseaver I checked this branch out locally and verified my changes. I can just send a commit to this branch if you're OK with it?

tseaver commented:
@dhermes Go ahead and push your changes.

dhermes committed:
Review changes.

- Got rid of trailing whitespace in README
- Made `mock` install conditional on Python 2 and added a `conftest.py`
  to monkey-patch it onto `unittest.mock`
- Compare against a `unittest.mock.call()`

dhermes committed:
Review changes.

- Using precise `typing` syntax for a `Generator[]` return type
- Using a `BinaryIO` for an argument type
- Adding a newline
- Putting a `b""` on the end of a stream

dhermes reviewed (approved):
LGTM

dhermes commented:
@tseaver Are we at a "stopping" point where I should re-build wheels?

tseaver commented:
> Are we at a "stopping" point where I should re-build wheels?

I think so.  I won't be able to tackle using the library in `google-resumable-media` / `google-cloud-storage` until next week at the earliest, but I think this library is at a good place to start that work.

dhermes commented:
OK great. Will file an issue and loop in the relevant folks.
